dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+import numpy as np
+import numpy.linalg as npl
+import scipy.ndimage as ndimage
+from dipy.align import floating
+import dipy.align.vector_fields as vf
+from dipy.align.mattes import MattesBase, sample_domain_regular
+from dipy.core.optimize import Optimizer
+from dipy.align.imwarp import (get_direction_and_spacings, ScaleSpace)
+from dipy.align.scalespace import IsotropicScaleSpace
+from dipy.core.optimize import SCIPY_LESS_0_12
+
+
+class MattesMIMetric(MattesBase):
+    def __init__(self, nbins=32, sampling_proportion=None):
+        r""" Initializes an instance of the Mattes MI metric
+
+        This class implements the methods required by Optimizer to drive the
+        registration process by making calls to the low level methods defined
+        in MattesBase.
+
+        Parameters
+        ----------
+        nbins : int, optional
+            the number of bins to be used for computing the intensity
+            histograms. The default is 32.
+        sampling_proportion : int in (0,100], optional
+            the percentage of voxels to be used for estimating the (joint and
+            marginal) intensity histograms. If None, dense sampling is used
+            (see notes regarding the difference between sparse and dense
+            sampling). The default is None.
+
+        Notes
+        -----
+        Since we use linear interpolation, images are not, in general,
+        differentiable at exact voxel coordinates, but they are differentiable
+        between voxel coordinates. When using sparse sampling, selected voxels
+        are slightly moved by adding a small random displacement within one
+        voxel to prevent sampling points to be located exactly at voxel
+        coordinates. When using dense sampling, this random displacement is
+        not applied.
+        """
+        super(MattesMIMetric, self).__init__(nbins)
+        self.sampling_proportion = sampling_proportion
+
+    def setup(self, transform, static, moving, static_grid2space=None,
```

Comment:
Hi @matthew-brett, thanks for taking the time to analyze this so comprehensively! regarding the `__init__` and `setup` methods, you are totally right, the metric is very light and it's not costly to make new instances. However I still think it is very useful and natural to have only one instance and use it to compare any pair of images we may need. For example, consider this test: https://github.com/nipy/dipy/pull/654/files#diff-5a512c70f85666010a0c8c3d13db8f3aR186, here the optimizer is the one that knows what images need to be compared, and it will use the provided instance whenever it needs to compute the value and gradient of the metric. Also, if we later have other metrics (Correlation Ratio, Normalized MI, global cross correlation, etc) the optimizer (`AffineRegistration`) will not need to create new instances of the metric, it will just use the provided/configured one. It's true that we may pass the class of the metric instead of the instance to `AffineRegistration` and let it do as many instances as it needs, but I think that may be more problematic in the future because, for example, the optimizer would need to know how to instantiate each metric, which limits us to define a unified way of instantiating the metrics (why not let the user instantiate the metric and do any needed configurations before passing it to the optimizer?).